load-gen/src/main/java/io/smartcat/data/loader/datasource/CSVDataSource.java

```diff
@@ -0,0 +1,75 @@
+package io.smartcat.data.loader.datasource;
+
+import io.smartcat.data.loader.api.DataSource;
+import org.apache.commons.csv.CSVParser;
+import org.apache.commons.csv.CSVRecord;
+
+import java.util.Iterator;
+
+/**
+ * Data source that reads data from CSV file.
+ *
+ * @param <T> Type of the data provided by data source implementation.
+ */
+public class CSVDataSource<T> implements DataSource<T>, AutoCloseable {
+
+    private final CSVParser csvParser;
+    private final RowMapper<T> rowMapper;
+
+    private Iterator<CSVRecord> iterator;
+    private boolean closed = false;
+
+    /**
+     * COnstructs CSV data source with specified <code>csvParser</code> and <code>rowMapper</code>.
+     * @param csvParser parser which provides CSV records to this data source.
+     * @param rowMapper mapper which maps CSV records to type {@code <T>}.
+     */
+    public CSVDataSource(CSVParser csvParser, RowMapper<T> rowMapper) {
+        if (csvParser == null) {
+            throw new IllegalArgumentException("CSV parser cannot be null.");
+        }
+        if (rowMapper == null) {
+            throw new IllegalArgumentException("Row mapper cannot be null.");
+        }
+        this.rowMapper = rowMapper;
+        this.csvParser = csvParser;
+        this.iterator = this.csvParser.iterator();
+    }
+
+    @Override
+    public boolean hasNext(long time) {
+        if (closed) {
+            throw new IllegalStateException("Data source is already closed.");
```

Comment:
Although a runtime exception is an appropriate response for `#getNext()` when data source is closed, I think here a simple `false` return value would be enough, especially because there is no explicit way to test 'closed' state.